renderdoc/driver/vulkan/vk_pixelhistory.cpp

```diff
@@ -1018,9 +1067,248 @@ struct TestsFailedCallback : public VulkanPixelHistoryCallback
+    VkPipelineMultisampleStateCreateInfo *ms =
+        (VkPipelineMultisampleStateCreateInfo *)ci.pMultisampleState;
+
+    // Only interested in a single sample.
+    ms->pSampleMask = &m_SampleMask;
+    // We are going to replay a draw multiple times, don't want to modify the
+    // depth value, not to influence later tests.
+    ds->depthWriteEnable = VK_FALSE;
+
+    if(pipeCreateFlags & PipelineCreationFlags_DisableCulling)
+      rs->cullMode = VK_CULL_MODE_NONE;
+    if(pipeCreateFlags & PipelineCreationFlags_DisableDepthTest)
+      ds->depthTestEnable = VK_FALSE;
+    if(pipeCreateFlags & PipelineCreationFlags_DisableStencilTest)
+      ds->stencilTestEnable = VK_FALSE;
+    if(pipeCreateFlags & PipelineCreationFlags_DisableDepthBoundsTest)
+      ds->depthBoundsTestEnable = VK_FALSE;
+
+    rdcarray<VkPipelineShaderStageCreateInfo> stages;
+    stages.resize(ci.stageCount);
+    memcpy(stages.data(), ci.pStages, ci.stageCount * sizeof(VkPipelineShaderStageCreateInfo));
+
+    for(size_t i = 0; i < ci.stageCount; i++)
+    {
+      if((ci.pStages[i].stage == VK_SHADER_STAGE_FRAGMENT_BIT) &&
+         (pipeCreateFlags & PipelineCreationFlags_FixedColorShader))
+      {
+        stages[i].module = m_ShaderCache->GetFixedColShader();
+        stages[i].pName = "main";
+      }
+      else if(replacementShaders[StageIndex(stages[i].stage)] != VK_NULL_HANDLE)
+      {
+        stages[i].module = replacementShaders[StageIndex(stages[i].stage)];
+      }
+    }
+    ci.pStages = stages.data();
+
+    if(!dynamicScissor)
+    {
+      VkRect2D *pScissors = (VkRect2D *)vs->pScissors;
+      for(uint32_t i = 0; i < vs->viewportCount; i++)
+      {
+        ScissorToPixel(vs->pViewports[i], pScissors[i]);
+        if(pipeCreateFlags & PipelineCreationFlags_IntersectOriginalScissor)
+          IntersectScissors(vs->pScissors[i], pScissors[i]);
+      }
+    }
+
+    VkPipeline pipe;
+    VkResult vkr = m_pDriver->vkCreateGraphicsPipelines(m_pDriver->GetDev(), VK_NULL_HANDLE, 1, &ci,
+                                                        NULL, &pipe);
+    RDCASSERTEQUAL(vkr, VK_SUCCESS);
+    m_PipeCache.insert(std::make_pair(pipeKey, pipe));
+    return pipe;
+  }
+
+  void ReplayDraw(VkCommandBuffer cmd, VkPipeline pipe, int eventId, uint32_t test)
+  {
+    m_pDriver->GetRenderState().graphics.pipeline = GetResID(pipe);
+    m_pDriver->GetRenderState().BindPipeline(cmd, VulkanRenderState::BindGraphics, false);
+
+    uint32_t index = (uint32_t)m_OcclusionQueries.size();
+    if(m_OcclusionQueries.find(rdcpair<uint32_t, uint32_t>(eventId, test)) != m_OcclusionQueries.end())
+      RDCERR("A query already exist for event id %u and test %u", eventId, test);
+    m_OcclusionQueries.insert(std::make_pair(rdcpair<uint32_t, uint32_t>(eventId, test), index));
+
+    ObjDisp(cmd)->CmdBeginQuery(Unwrap(cmd), m_OcclusionPool, index, VK_QUERY_CONTROL_PRECISE_BIT);
```

Comment:
We shouldn't ever need the precise result I think, since we only ever want to know if zero or non-zero samples passed. Since this comes with a physical device feature being needed, could we remove this flag?.

Reply:
You are right, changed.